### PR TITLE
improvement: changes to internal RPC Quantity type to make basic mathematic operations cleaner

### PR DESCRIPTION
**Pull request by jeffsmale90**
Add addition and multiplication methods  to internal RPC-JSON `Quantity` data type.

This change impacts Ganache internals only, except for some minor performance improvements.

### CODE REVIEW

**Comment by jeffsmale90** [comment]
This change went back and forward a little bit - initially I implemented buffer level addition in `buffer-math.ts`, which was somewhat promising, but ultimately I believe that the minor performance improvements that we gained were not worth the increased complexity - in the case of adding a `number`, `Buffer`, or `Quantity` type, this was up to ~10% faster, and 1-2% faster (wooo) if adding a `bigint`.

I deleted those changes, but left them in the history for anyone who is curious, or have better ideas about how we could approach this.

https://github.com/trufflesuite/ganache/commit/4cad19352e19e2884c91d3ccc0783fccc068df03 is changes to consumers of `.add()` and `.multiply()`, plus a couple of changes to use the static `Quantity.toXXX()` methods (which I'm happy to remove if anyone feels they don't belong here. [View just the first 6 commits](https://github.com/trufflesuite/ganache/pull/3164/files/de1a6e545e23b5e31cd7a0e26defc3e9b31bd6d2) to see the changes in isolation to the consumer changes.

Oh and I used property-based testing with `fast-check`. I might look to expand this to the rest of the `JSON-RPC` tests at some stage.

File: src/chains/ethereum/ethereum/src/api.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/api.ts]:
```suggestion
        parentHeader.number.add(1n),
```

File: src/chains/ethereum/ethereum/src/blockchain.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/blockchain.ts]: `BN` actually allows a `Buffer` in its constructor. I wonder if we should change it.

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]: Can we do subtraction?, i.e., `blockNumber.add(-1n)`?

**Comment by jeffsmale90** [comment]
This has become funky after rebasing the upstream branch, so I will have to address that in the morning.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (MicaiahReid) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: Do we need to do any verification that the characters are only hexadecimal, or is this safe because it's for internal use only?

Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: We rely on `BigInt(value)`'s validation here. We have to explicitly check the first two chars, otherwise it may parse as decimal. I've added a few more test cases: https://github.com/trufflesuite/ganache/pull/3164/commits/e5b747bc9e6abb3858b2d098c7afd14bfb916ee8

**Review by MicaiahReid — APPROVED** [review]

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (MicaiahReid) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: Ahh there already were tests with invalid hex, so you probably didn't need to add those. My miss!

**Review by davidmurdoch — CHANGES_REQUESTED** [review]
just a quick partial review

File: src/chains/ethereum/ethereum/src/blockchain.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/blockchain.ts]: Yeah, I like that, it's cleaner and it's less work.

File: src/chains/ethereum/ethereum/src/api.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/api.ts]: Done!

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]: We do not presently support subtraction. I intentionally did not support it, as (IIRC) there's only one case of this, and it simplifies the implementation to use the same parsing / validation logic as the Quantity constructor (which doesn't support negatives).

We _could_ implement this, but I'm not sure whether it's worth the effort. Happy to hear differing opinion though.

**Comment by jeffsmale90** [comment]
There's a bunch of conflicts on here, I'll rebase to develop and resolve those tomorrow.

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]: Actually it's not that hard, I implemented support for -ve addend (so long as the sum >= 0). But only support 
 for `number | bigint` type addends (as other types don't have a clear way to represent negative numbers).

And I found an issue when testing this! I opened #3328 to fix an issue with converting buffer to number (I put a fix in this PR also, but I'll resolve any divergence when I merge).

As commented on the related PR, I think I'll convert some more of the tests in here into FC tests after I've merged a couple of the outstanding tests in this area.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:211]: I added overrides to the `toXXXX()` methods so that calling the function without supplying `nullable` (defaults to `false`), returns a non-nullable type. It'd be nice to do this on the instance methods too, but I think that will be a more significant undertaking.

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]: As commented on the related PR, I think I'll convert some more of the tests in here into FC tests after I've merged a couple of the outstanding tests in this area.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:242]: I added overrides to the `toXXXX()` methods so that calling the function `nullable: false`, returns a non-nullable type. It'd be nice to do this on the instance methods too, but I think that will be a more significant undertaking, that I'll pick up outside of here.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: This check will fail if the `Quantity` object passed in is not an instance of _this_ `Quantity` class. We should probably check for `"toBigInt" in argument` or something like that.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: If you use a `switch` as `switch (typeof argument) {` you won't have to `as number`, etc, as typescript will narrow the type automatically:

```
switch (typeof argument) {
  case "number":
    if (argument % 1) {
       // ...
    break;
  case "string":
    // ...
  default:
    if (Buffer.isBuffer(argument)) //... etc
}
```

File: src/chains/ethereum/ethereum/src/transaction-pool.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/transaction-pool.ts:220]: This new approach will much more expensive because each intermediate step gets converted back to a Buffer.

I worry these new methods are too expensive to be so easy to use. haha

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]:
```suggestion
              blockNumber.add(-1n).toString()
```

File: src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/api/eth/getCode.test.ts]:
```suggestion
            const nextBlockNumber = Quantity.from(await provider.send("eth_blockNumber")).add(1n);
```

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:263]: worth updating this to the new static Quantity helper in the case that `value` is _not_ null?

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]:
```suggestion
    if (multiplierValue < 0n) {
```

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:219]: Why do we need to guard against this case? `getArgumentAsBigInt` can't return `undefined`, right?

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: Why do we need to guard against this case? `getArgumentAsBigInt` can't return `undefined`, right?

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]:
```suggestion
      throw new Error(`Cannot add ${addend} to a Quantity of ${this}, as it results in a negative value`);
```

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: It doesn't look like this is used. Am I wrong?

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: This check will fail if the `Quantity` object passed in is not an instance of this `Quantity` class. We should probably check for `"toBigInt" in argument` or something like that.

**Review by davidmurdoch — CHANGES_REQUESTED** [review]
partial review I wanted to get over real quick!

File: src/chains/ethereum/ethereum/src/transaction-pool.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/transaction-pool.ts:220]: I've added 31161d0c29ca0ac93bd8ed0dc7ba116effa0fd5a which defers conversion to `Buffer` and will use the raw numeric value where possible.

This means that `Quantity.from(1).toBigInt()` will never convert to `Buffer`, as will the code that you linked.

Keen to hear your thoughts.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: I've never been a huge fan of switch statements, but I guess they're no worse than big if / else chains. Also there's obvious benefits in cases like this in TS.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: It's also effectively the same as `Quantity.Empty`

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: Agreed - although I think that's a bug!

I've changed it to not coerce invalid types to 0, and moved the error into the getArgumentAsBigInt function.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:219]: Same as below.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts]: I <3 `switch` statements (they're usually faster, too)

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:162]: looks like these files all need to be prettiered

Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:162]: Do this:

```suggestion
    let argumentValue: bigint;

    switch (typeof argument) {
```

and then remove the `as number`s below.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:187]: this can't happen (per the types for `argument`)

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:263]: Clarification: I meant  `Quantity.getArgumentAsBigInt(value)` when we know `value != null`

File: src/packages/utils/package.json
Comment (davidmurdoch) [inline src/packages/utils/package.json:59]: This should be pinned to an exact version.

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/input-parsers.ts:38]: This doesn't return a `Buffer` anymore.

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/input-parsers.ts:55]: `input` is already a `number`

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/input-parsers.ts:72]: `input` is already a bigint

File: src/chains/ethereum/ethereum/src/transaction-pool.ts
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/src/transaction-pool.ts:220]: `31161d0` reminds me a lot of the original `Quantity`/`Data` implementation (though this way is simpler as it only ). I'm very curious about the performance here.

File: src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts:62]: Any reasons not to use this here:

```suggestion
    return this.toBuffer();
```

File: src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts:70]: Any reason not to use this here:


```suggestion
    return this.toBuffer() == null;
```

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:219]: We still don't need to guard against it, right?

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:163]: Can you run a test to determine the best order for these `switch` cases? They should be ordered from mostly likely to be used to least likely.

File: src/chains/ethereum/ethereum/src/api.ts
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/api.ts:927]: Since we `add(1n)` more than anything else I wonder if a highly optimized `.increment()` function would be worth while. Maybe we do all the Math in a Buffer directly or something. :thinking: (this probably can/should be saved for another PR)

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:19]: This actually can't happen (per the types). Can we remove this check altogether?

```suggestion
```

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (davidmurdoch) [inline src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:234]: This is used in two places, and i've suggested we remove it in one of them, leaving us with the `toBigInt` check. Which means we really only need to check if the thing passed in has a `toBigInt` function, as that's all the object really needs anyway. Checking just the one property would definitely be faster than checking all of these methods.

`"toBigInt" in value && typeof value.toBigInt === "function` could replace `Quantity.isQuantity(value)`.

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: src/chains/ethereum/ethereum/src/api.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/api.ts:927]: I'll look into that. Initially I did all of the math directly in the buffer, and got somewhat promising results, but I considered the complexity to be unreasonable for the minimal gains. It is likely much more significant for a simple increment() function however, as the conversion to bigint and back is a more singificant percentage of the entire operation. I will look into it.

Edit: I created #3390 to cover this.

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (MicaiahReid) [inline src/packages/utils/src/things/json-rpc/input-parsers.ts:8]: Is there any difference between `private _resolver` and `#resolver`? If not, we should switch to be consistent with the rest of the repo.

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (MicaiahReid) [inline src/packages/utils/src/things/json-rpc/input-parsers.ts:7]: I think a js-doc comment explaining this class would be worthwhile

File: src/packages/utils/src/things/json-rpc/input-parsers.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/input-parsers.ts:8]: @davidmurdoch has pointed out that # prefixed private class fields perform significantly worse, in part due to how webpack polyfills these (I think?)

I don't know whether there's a reasonable fix to this, but I've updated #3240 as a tech-debt to cover this issue.

File: src/chains/ethereum/ethereum/src/transaction-pool.ts
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/transaction-pool.ts:220]: Right? I had flashbacks.

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:219]: Correct!

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:234]: That's fair enough. I think I need to get out of thinking of these as part of a public api...

File: src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts
Comment (jeffsmale90) [reply src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts:19]: That's true, but we do use this in the `persistent-cache` key generation. 

Blocknumber is passed in as either string or Quantity instance. I've pushed that differentiation back out into `persistent-cache`, so that the `Quantity` implementation is cleaner.

**Comment by jeffsmale90** [comment]
I've done some performance analysis on this, and created the benchmark suite here: https://github.com/jeffsmale90/quantity-perf

I think that @davidmurdoch 's concern is correct - adding these helper methods make the performance of chained math functions very poor. The work arounds I have added show some promise, but are adding complexity (maybe there's something in there that could be used to benefit Quantity performance generally).

I believe that this is no longer a valuable thing to pursue, so will put this on hold, and may return to it at a later date.

**Pull request closed without merging**
